Provides support for 'allowed_files' configuration

rthbound opened the pull request:
Provides support for 'allowed_files' configuration
  - As an alternative to allowing certain incantations of offensive terms,
it may be desirable to limit the occurrence of some certain term to
specific files only.
  - Use case for consideration: A project relies on some library. The
library requires a configuration file which utilizes an offensive
term. The project developers wish to disallow the offensive term but
also wish to make an exception for the aforementioned library's
configuration file.

Extra:
- This commit addresses rubocop offenses (most, not all)
- This commit fixes a previously undetected bug, where inclusive-code
  would fail to find offenses within filenames if the flagged terms
  configuration specified zero allowed patterns

rthbound committed:
Provides support for 'allowed_files' configuration

  - As an alternative to allowing certain incantations of offensive terms,
    it may be desirable to limit the occurrence of some certain term to
    specific files only.
  - Use case for consideration: A project relies on some library. The
    library requires a configuration file which utilizes an offensive
    term. The project developers wish to disallow the offensive term but
    also wish to make an exception for the aforementioned library's
    configuration file.

  Extra:
    - This commit addresses rubocop offenses (most, not all)
    - This commit fixes a previously undetected bug, where inclusive-code
      would fail to find offenses within filenames if the flagged terms
      configuration specified zero allowed patterns

helenuria commented on ruby/spec/inclusive_code/cop/inclusive_code_spec.rb:
```diff
@@ -92,6 +92,71 @@ class MyInclusiveEngine
+        described_class.new(nil, nil, flagged_terms)
+      end
+
+      it 'can ignore offenses within the path of an allowed file while detecting the'\
```
```suggestion
      it 'ignores offenses in the allowed file and does not ignore them in another file. ' do
```

helenuria commented on ruby/spec/inclusive_code/cop/inclusive_code_spec.rb:
```diff
@@ -92,6 +92,71 @@ class MyInclusiveEngine
+      end
+
+      it 'can ignore offenses within the path of an allowed file while detecting the'\
+        'same offenses in the path of another file' do
```
```suggestion
```

rthbound replied:
Just wanted to share that you can leave a multi-line suggestion by (while holding shift) clicking the top line then the bottom line to select a range of lines. From there you click to add a comment and insert the suggestion as usual.

helenuria commented on ruby/spec/inclusive_code/cop/inclusive_code_spec.rb:
```diff
@@ -92,6 +92,71 @@ class MyInclusiveEngine
+        expect_no_offenses(source.lines.first, 'lib/rubocop/cop/inclusive_code.rb')
+      end
+
+      it 'can ignore offenses in one file while detecting other offenses in the same file' do
```
```suggestion
      it 'ignores allowed offenses in an allowed file while detecting offenses in the same file' do
```

helenuria commented on ruby/spec/inclusive_code/cop/inclusive_code_spec.rb:
```diff
@@ -92,6 +92,71 @@ class MyInclusiveEngine
+        expect_offense(source, 'README.md')
+      end
+
+      it 'does not detect offenses in an allowed filename' do
```
this seems redundant with the the first test

rthbound replied:
Good eye, thank you

helenuria reviewed (approved):
I really like this feature, its much more fine tuned than excluding whole files in rubocop.yml. Added some suggestions to the spec, but feel free to ignore.

rthbound commented:
Thank you for all the suggestions. I'll take a closer look this evening.

I'm glad it somehow offers more than rubocop itself -- I'm not extremely familiar with rubocop's features 😬

rthbound committed:
Simplifies wording in test scenario titles

Also removes a redundant test.

Co-authored-by: Helen Stockman <helen.n.stockman@gmail.com>

helenuria reviewed (approved):
🎉

helenuria commented:
Also, it'd be great to document this feature in the ruby directory's README

rthbound committed:
Adds README documentation for some of the supported configurations

rthbound commented on ruby/README.md:
```diff
@@ -29,8 +29,75 @@ Flexport/InclusiveCode:
 You can run the cop on a specific file with `rubocop --only Flexport/InclusiveCode file_path`.
 
 If you want to add inline `rubocop:disable` or `rubocop:todo` comments on all offenses, set `DisableAutoCorrect: true` in your .rubocop.yml, and run `rubocop --only Flexport/InclusiveCode --auto-correct --disable-uncorrectable`.
+
```
@helenuria, here's my attempt at expanding the README to include more details about configuration.

What do you think?

I think it would be good to indicate which key/value pairs are required and which are optional. Could you help me by providing that information?

Some specific questions that came up for me when writing this:
- is it required to provide suggestions?
- if no suggestions are provided, will autocorrect substitute an empty string for any offenses it finds?

rthbound replied:
Adding specs for my questions... It appears that:
- ~it is not required to provide suggestions~
- if no suggestions are provided, auto-correct would work as is likely desired (i.e. it would make no change)

Currently we do require a value for the `allowed` key. It would probably preferable to accept a flagged terms config having no `allowed` key.

I'll update specs to ~reflect~ explore some of these ~findings~ questions, and then take a look at including any relevant info in the readme.

helenuria replied:
I agree itd be preferable to accept flagged terms config having no allowed key.

rthbound committed:
Determines minimal configuration for a flagged term (no suggestions, no allowed exceptions)

- Updates library to support a configuration having no suggestions
- Updates library to support a configuration having no allowed exceptions
- Adds test for a configuration having no suggestions
- Adds test for a configuration having no allowed exceptions
- Updates README.md with more details and sample configurations

helenuria commented on ruby/README.md:
```diff
@@ -29,8 +29,75 @@ Flexport/InclusiveCode:
+flagged_terms:
+  whitelist:
+    suggestions:
+      - main
```
huge nit but `allowlist` not `main`

helenuria commented on ruby/README.md:
```diff
@@ -29,8 +29,75 @@ Flexport/InclusiveCode:
+
+### Suggestions and Autocorrect
+
+In a document titled [Terminology, Power and Offensive Language](https://tools.ietf.org/id/draft-knodel-terminology-01.html), the Internet Engineering Task Force (IETF) recommends that an editor or reviewer *should* "offer alternatives for offensive terminology as an important act of correcting larger editorial issues and clarifying technical concepts."
```
nice research!

rthbound replied:
thank you! I really like that document (_their_ research, _their_ arguments)

helenuria commented:
new commit LGTM 👍
